### ISSUE TITLE
use mixFilter to create a stereo track from 2 mono tracks

### ISSUE THREAD

**Issue by ackjewtn** [kaltura/nginx-vod-module#1139]
As the title explains, i've been trying to mix a1 (mono) and a2 (mono) to become 1 stereo track.

Using the following JSON:
`{
  "sequences": [
    {
      "id": "video",
      "clips": [
        {
          "type": "source",
          "path": "/opt/static/videos/test_clip.prx.mp4",
          "tracks": "v1"
        }
      ]
    },
    {
      "id": "audio",
      "clips": [
        {
          "type": "mixFilter",
          "sources": [
            {
              "type": "source",
              "path": "/opt/static/videos/test_clip.prx.mp4",
              "tracks": "a1"
            },
            {
              "type": "source",
              "path": "/opt/static/videos/test_clip.prx.mp4",
              "tracks": "a2"
            }
          ]
        }
      ]
    }
  ]
}`

Source file:
`
    Stream #0:0(eng): Video: h264 (Main) (avc1 / 0x31637661), yuv420p(tv, bt709), 1280x720 [SAR 1:1 DAR 16:9], 1495 kb/s, 25 fps, 25 tbr, 2500 tbn, 50 tbc (default)
    Metadata:
      encoder         : AVC Coding
      timecode        : 01:13:07:24
    Stream #0:1(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, mono, fltp, 61 kb/s (default)
    Stream #0:2(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, mono, fltp, 61 kb/s (default)
    Stream #0:3(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, mono, fltp, 61 kb/s (default)
    Stream #0:4(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, mono, fltp, 61 kb/s (default)
    Stream #0:5(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, mono, fltp, 61 kb/s (default)
    Stream #0:6(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, mono, fltp, 61 kb/s (default)
    Stream #0:7(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, mono, fltp, 61 kb/s (default)
    Stream #0:8(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, mono, fltp, 61 kb/s (default)
`

Should this be possible? I still get a mono track presented in the HLS playlist.

`
#EXTM3U

#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="audio0",NAME="default",AUTOSELECT=YES,DEFAULT=YES,CHANNELS="1",URI="index-saudio-a1.m3u8"

#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=1556956,RESOLUTION=1280x720,FRAME-RATE=25.000,CODECS="avc1.4d401f,mp4a.40.2",VIDEO-RANGE=SDR,AUDIO="audio0"
index-svideo-v1.m3u8
`

**Comment by erankor** [comment]
No, this is not supported. The mix filter blends together multiple audio track, for example, you can merge background music with audio description (it maps to the `amix` ffmpeg filter).
It's probably not too hard to add another filter that will do what you need (using the ffmpeg `amerge` filter), but unfortunately, I'm not available for this...

**Comment by ackjewtn** [comment]
Thanks for the response @erankor 

I actually took a quick look at `mix_filter.c` yesterday and saw the same. I also tried to just change `MIX_FILTER_DESC_PATTERN` from `amix` to `amerge` since the filter pattern should be more or less identical and recompiled but that didn't change anything. I'm not very familiar with the code, so i'm struggling a bit to understand how the arguments are being passed to the filter, but since this is done in ffmpeg by doing `-filter_complex "amerge=inputs=2[audio]" -map "[audio]"` i thought it would work.

What is the preferred way of adding `amerge` as a filter? Implementing it inside `mixFilter`, or adding another filter, e.g. `mergeFilter`?

**Comment by erankor** [comment]
The problem is probably because the module uses one of the input tracks to determine the media info of the output track - and specifically what you care about here - the number of channels. In other words, the module will use one of your mono tracks as reference, and will tell ffmpeg to generate mono output. So, what probably happens is that the two mono tracks are merged into stereo, but then converted back to mono by the filter sink (I don't know if it mixes them / uses only one of them etc.).
It will take some refactor of the audio filtering to do this correctly, maybe each filter would return the media info of its output. So for example, the mix filter will use the media info of the first track, while the merge filter will count the total number of channels and use that. Actually, that's more work than I had in mind when I commented above...
Maybe you can hack it for your specific use case, for example, add something like - if (top level filter is merge) { sum the number of channels and use it for the output media info }
